Add `ImageDecoder`-independent animated AVIF test

mathiasbynens opened the pull request:
Ref. https://github.com/web-platform-tests/interop/issues/136

mathiasbynens committed:
Add `ImageDecoder`-independent animated AVIF test

Ref. https://github.com/web-platform-tests/interop/issues/136

mathiasbynens commented on animated-avif/animated-avif-timeout.html:
```diff
@@ -0,0 +1,11 @@
+<html class="reftest-wait">
+<title>Animated AVIF: Second frame displays quickly, replacing red with green.</title>
+<link rel="match" href="animated-avif-timeout-ref.html"/>
+<img src=../images/animated-avif.avif onload="loaded()"/>
```
Note: There might be a better way to create this image or make it smaller. The way I did it feels a bit hacky:

```sh
ffmpeg -loglevel error -hide_banner -f lavfi -i color=red:2x2 -frames:v 1 -y /tmp/red.png;
ffmpeg -loglevel error -hide_banner -f lavfi -i color=green:2x2 -frames:v 1 -y /tmp/green.png;
ffmpeg -loglevel error -hide_banner -i "concat:/tmp/red.png|$(node -p '"|/tmp/green.png".repeat(30*60)')" -y /tmp/test.avif;
```

foolip replied:
Hmm, a lower framerate would definitely be better. Seems like this ought to be an option for lavfi rather than a filter? https://trac.ffmpeg.org/ticket/5945 seems releavant.

Maybe the APNG test file can just be transcoded?

mathiasbynens replied:
@vigneshvg might know a better solution.

vigneshvg replied:
i'm sorry i don't understand what the desired output here is.

do we want an animated AVIF with 30 red frames followed by 30 green ones?

mathiasbynens replied:
The test runner checks the rendered image after roughly 1 second and verifies that it sees a green frame as opposed to a red frame. In that case, animated AVIF must be supported. So we want either:

- 1 red frame followed by lots of green ones (the above example uses 60 seconds worth of green frames, but 10 seconds might be sufficient), or
- 1 short red frame followed by a loooooooooong green frame (if AVIF supports variable framerate)

vigneshvg replied:
ah okay. it is perhaps possible with ffmpeg, but for a quick solution you can do this very easily with avifenc.

ffmpeg -loglevel error -hide_banner -f lavfi -i color=red:2x2 -frames:v 1 -y /tmp/red.png;
ffmpeg -loglevel error -hide_banner -f lavfi -i color=green:2x2 -frames:v 1 -y /tmp/green.png;
avifenc --duration 1 /tmp/red.png --duration 1000 /tmp/green.png /tmp/anim.avif

The second duration parameter is what controls the duration of the second frame.

A sample data URL of such an AVIF file: https://paste.googleplex.com/5458439613448192

mathiasbynens replied:
Thanks! I’ve used this in https://github.com/web-platform-tests/wpt/commit/6f2cab7549adae38c7bfbcd9b8e92d48568d128b. I’ve also recreated `green.avif` using a similar command (mentioned in the commit message) but even then successful test runs in Chrome show a color value difference of up to 3 per channel, so I added [`<meta name="fuzzy" content="maxDifference=3">`](https://web-platform-tests.org/writing-tests/reftests.html#fuzzy-matching). It would be great to avoid this but I don’t really know where the difference is coming from.

mathiasbynens replied:
I’ve reverted the optimization for now since it doesn’t feel right to have to introduce a fuzziness factor. Let’s investigate this in a separate follow-up PR?

vigneshvg replied:
> but even then successful test runs in Chrome show a color value difference of up to 3 per channel

It is probably the compression artifacts as avifenc is lossy by default. Can you try passing `--lossless` to `avifenc` as the first parameter ?

mathiasbynens replied:
https://github.com/web-platform-tests/wpt/pull/37321

mathiasbynens committed:
Resolve lint issue

mathiasbynens committed:
Use AVIF version of green ref image

mathiasbynens committed:
Restore `setTimeout`

Using `step_timeout` instead, as recommended by the linter, makes the test time out.

mathiasbynens committed:
Attempt to directly create the green ref image

```
ffmpeg -loglevel error -hide_banner -f lavfi -i color=green:2x2 -frames:v 1 -y /tmp/green.avif
```

mathiasbynens committed:
Optimize AVIF assets

The `animated-avif.avif` and `green.avif` files are generated as follows:

```sh
ffmpeg -loglevel error -hide_banner -f lavfi -i color=red:2x2 -frames:v 1 -y /tmp/red.png;
ffmpeg -loglevel error -hide_banner -f lavfi -i color=green:2x2 -frames:v 1 -y /tmp/green.png;
avifenc --duration 1 /tmp/red.png --duration 1800 /tmp/green.png /tmp/animated-avif.avif;
avifenc --duration 1 /tmp/green.png /tmp/green.avif;
```

mathiasbynens committed:
Revert "Optimize AVIF assets"

This reverts commit d1fcc1a49729d79c71681ea87e30d13140cea72a.

foolip commented:
https://wpt.fyi/results/animated-avif/animated-avif-timeout.html?label=pr_head&max-count=1&pr=36044 shows a failure for Firefox. With only 4 pixels being different, it can't be that the red frame is still showing. Can you check what's going on?

mathiasbynens committed:
Rename directory

mathiasbynens commented on animated-avif/META.yml:
```diff
@@ -0,0 +1 @@
+spec: https://aomediacodec.github.io/av1-avif/
```
Done.

mathiasbynens commented:
> https://wpt.fyi/results/animated-avif/animated-avif-timeout.html?label=pr_head&max-count=1&pr=36044 shows a failure for Firefox. With only 4 pixels being different, it can't be that the red frame is still showing. Can you check what's going on?

Why not? The image is 2×2 px, Firefox doesn’t support animated AVIF yet, so I would expect the red frame to show.

foolip commented:
Aha, didn't know it was a 2x2 image. I assumed a 100x100 or something.
